### PR TITLE
Rebase #36720 "Add exec option to tmpfs"

### LINKED ISSUE #32131: Swarm mode, mount tmpfs, how to remove noexec?
**Description**

Swarm mode, ```docker service create --mount type=tmpfs```,  how to remove noexec flag?

**Steps to reproduce the issue:**
1.  docker service create --mount type=tmpfs,destination=/data,tmpfs-size=100M -p:22 rastasheep/ubuntu-sshd:14.04
2. ssh root@localhost -p 30000
3. mount | grep /data

**Describe the results you received:**
tmpfs on /data type tmpfs (rw,nosuid,nodev,noexec,relatime,size=102400k)

**Describe the results you expected:**
I want to remove noexec flag, how? like, ```docker run -it --rm --tmpfs /data:exec bash```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:04 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:53:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 7
Server Version: 17.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 72
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: 8t1tc9olkkkvto0m9ap20yg4q
 Is Manager: true
 ClusterID: mjf13s8g3luwoo579kbtkxvau
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 172.25.20.22
 Manager Addresses:
  172.25.20.22:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-98-generic
Operating System: Ubuntu precise (12.04.5 LTS)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 993.5 MiB
Name: docker0
ID: HPTL:DPNP:QNJ6:6K5R:NFNT:CMFE:RWWP:TWAW:5DAO:LZXV:DEKY:5VKZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 44
 Goroutines: 136
 System Time: 2017-03-27T10:03:52.81516827Z
 EventsListeners: 1
Username: huozic
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.): **
VirtualBox 5.1.16 r113841
ubuntu 12.04 3.13.0-98-generic

### LINKED ISSUE #45385: Add `--mount` tmpfs with `exec` support for vscode devcontainer
### Description

vscode devcontainer.json
```
  "mounts": [
    "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec=true",
```

```
invalid argument "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec=true" for "--mount" flag: unexpected key 'exec' in 'exec=true'
See 'docker run --help'.
```

---

vscode devcontainer.json
```
  "mounts": [
    "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec",
```

```
invalid argument "type=tmpfs,destination=/tmp,tmpfs-mode=1777,exec" for "--mount" flag: invalid field 'exec' must be a key=value pair
See 'docker run --help'.
```

---

https://github.com/moby/moby/issues/32131#issuecomment-1031359653

https://github.com/moby/sys/pull/66

That pr is merged, but what happend here?

### PR DESCRIPTION
**Pull request by dperny**
addresses:

- fixes https://github.com/moby/moby/issues/32131
- fixes https://github.com/moby/moby/issues/45385

I'm eliding the typical format for PRs because this PR is simply a rebase of an existing PR, #36720 onto the current master.

Closes #36720.

### CODE REVIEW

File: api/swagger.yaml
Comment (thaJeztah) [inline api/swagger.yaml:439]: ISTR there was some discussion whether this should be a `string`, a `[]string`  (or a `map[string]string`, but I think we didn't want a map), did we ever reach full consensus on that?

File: api/types/mount/mount.go
Comment (cpuguy83) [inline api/types/mount/mount.go:121]: It seems like we only plan to have a very small subset of options here.
Does it make sense to promote those optioons to fields on this struct?
The ones we have are difficult to represent accurately, I admit (default exec vs explicit exec vs noexec).

UID/GID are probably worth adding at some point... if someone asks for it (this may have already happened).
I'm not sure I agree with the comment about implications in the cluster.

I'm not sure the other options would be worthwhile.

**Review by cpuguy83 — COMMENTED** [review]
I'm going to be a little pedantic on having a TmpfsOptions struct with a generic `Options` field inside of it.
I don't want to block this on pedantry, especially since noexec is a good thing to add so happy to bow out if others think this is fine.

**Comment by corhere** [comment]
We discussed this during today's maintainer call. The main sticking point is that the daemon is tokenizing and parsing the options string. Parsing would not be needed if the options were encoded as structured data in the API request. Nor would parsing be needed if the daemon didn't validate the options against an allowlist: it would simply [concatenate the user options to the end of the daemon-generated options string](https://github.com/opencontainers/runc/issues/3911#issuecomment-1603194295) and pass that along to the kernel. No matter which decision we make on validating the options against an allowlist, there is no need for the daemon to parse a mount options string.

Consensus was in favour of keeping validation in, based on the "yes is forever, no is temporary" principle of API design. We can always expand the allowlist or remove the validation in the future, but we can't walk it back. (We would also like to add a few more obviously safe options to the allowlist, such as `nosuid`, `nosymfollow`, `dev`/`nodev`, `suid`/`nosuid`.) That means the options need to be encoded as structured data in the API so the daemon can validate them without parsing. For future-proofing, I would ideally like the structure to map cleanly onto [the new `fsconfig` syscall](https://elixir.bootlin.com/linux/v6.8.7/source/fs/fsopen.c#L307) API.
- Key and value are separate in the structured data
- Ordered list, in case the order that parameters are supplied may be significant

The outer data structure has to be a JSON Array. The only remaining question is how to encode individual parameters. At the very least we have to distinguish between flags and string parameters in the API as some flags will not be accepted through the `fsconfig_set_string` command.

If we limit ourselves to making only `fsconfig_set_flag` and `fsconfig_set_string` parameters representable in the API, it could be as simple as encoding flags as bare strings and string parameters as a 2-tuple of (string, string), encoded in JSON as a two-element array.
```json
[ "noexec", ["uid", "12345"] ]
```
Or use 2-tuples consistently, but use `null` in the value position to signify a flag?
```json
[ ["noexec", null], ["uid", "12345"] ]
```
If they're all 2-tuples, they could simply be flattened into an array where even indices are keys and odd indices are values.
```json
["noexec", null, "uid", "12345"]
```

If we really want to future-proof for the other kinds of parameter, we could (also?) accept a more verbose struct (JSON Object) representation, e.g.:
```json
[
  {"kind": "flag", "key": "noexec"},
  {"kind": "string", "key": "uid", "value": "12345"},
  {"kind": "binary-base64", "key": "some-binary-blob", "value": "eQo="},
  {"kind": "host-path", "key": "lowerdir", "value": "/path/to/thing"}
]
```
Or only be verbose in the value position.
```json
[
  "noexec", null,
  "uid", "12345",
  "gid": {"kind": "string", "value": "23456"},
  "rw", {"kind": "flag"},
  "some-binary-blob": {"kind": "binary", "content-encoding": "base64", "value": "eQo="}
]
```

The tmpfs filesystem currently only accepts flag and string parameters (and no in-tree filesystem accepts binary parameters as of Linux v6.8) so we don't need to implement support for the other kinds yet. We only need to ensure that whatever shape we choose can afford being extended to support expressing the other kinds of parameter when the time comes.

--------

We also considered adding the mount options to the outer `Mount` struct, but have decided not to pursue that for now. We could add one in the future without conflicting with the `TmpfsOptions` mount options, so long as we concatenate the options (or apply the fsconfig) from the different sources in a consistent order.

**Comment by cpuguy83** [comment]
```
volume/mounts/linux_parser_test.go:257:54: cannot use "exec" (untyped string constant) as [][]string value in struct literal
volume/mounts/linux_parser_test.go:263:37: cannot use "INVALID" (untyped string constant) as [][]string value in struct literal
```

**Review by thaJeztah — DISMISSED** [review]
LGTM

File: daemon/cluster/convert/container.go
Comment (corhere) [inline daemon/cluster/convert/container.go:598]: You have to explicitly discard the error to appease the linter.
```suggestion
	_ = json.Unmarshal([]byte(options), &unstring)
```

File: daemon/cluster/executor/container/container.go
Comment (corhere) [inline daemon/cluster/executor/container/container.go:367]:
```suggestion
		_ = json.Unmarshal([]byte(m.TmpfsOptions.Options), &options)
```

**Review by corhere — APPROVED** [review]

**Comment by thaJeztah** [comment]
Looks like something is failing in swagger validation;

```
Error: gger.yaml:452:15: [error] syntax error: expected <block end>, but found '<block mapping start>' (syntax)
Warning: er.yaml:2306:7: [warning] comment not indented like content (comments-indentation)
make: *** [Makefile:249: validate-swagger] Error 1
Error: Process completed with exit code 2.
```

File: docs/api/version-history.md
Comment (thaJeztah) [inline docs/api/version-history.md:34]: Does `/services/{id}/update` also accept it? (we can amend in a follow-up if needed)

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**
